// graphql-compose-pagination/lib/types/paginationInfoType.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _graphqlCompose = require('graphql-compose');

var GraphQLBoolean = _graphqlCompose.graphql.GraphQLBoolean,
    GraphQLObjectType = _graphqlCompose.graphql.GraphQLObjectType,
    GraphQLNonNull = _graphqlCompose.graphql.GraphQLNonNull,
    GraphQLInt = _graphqlCompose.graphql.GraphQLInt;


var PaginationInfoType = new GraphQLObjectType({
  name: 'PaginationInfo',
  description: 'Information about pagination.',
  fields: function fields() {
    return {
      currentPage: {
        type: new GraphQLNonNull(GraphQLInt),
        description: 'Current page number'
      },
      perPage: {
        type: new GraphQLNonNull(GraphQLInt),
        description: 'Number of items per page'
      },
      pageCount: {
        type: GraphQLInt,
        description: 'Total number of pages'
      },
      itemCount: {
        type: GraphQLInt,
        description: 'Total number of items'
      },
      hasNextPage: {
        type: GraphQLBoolean,
        description: 'When paginating forwards, are there more items?'
      },
      hasPreviousPage: {
        type: GraphQLBoolean,
        description: 'When paginating backwards, are there more items?'
      }
    };
  }
});

exports.default = PaginationInfoType;
